### PR TITLE
HIVE-26055: Added the right HivePrivilegesObjects for Alter table ren…

### PR DESCRIPTION
**Pull request by saihemanth-cloudera**
…ame command

### What changes were proposed in this pull request?
Added the right  HivePrivilegesObjects for Alter table rename command.

### Why are the changes needed?
So that authorization happens correctly in ranger/sentry

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Manual testing, Added unit tests.

### CODE REVIEW

File: itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOutputTableLocationSchemeIsFileHook.java [hunk 1]
```diff
@@ -24,10 +24,10 @@ public class VerifyOutputTableLocationSchemeIsFileHook implements ExecuteWithHoo
 
   @Override
   public void run(HookContext hookContext) {
-    for (WriteEntity output : hookContext.getOutputs()) {
-      if (output.getType() == WriteEntity.Type.TABLE) {
-        String scheme = output.getTable().getDataLocation().toUri().getScheme();
-        Assert.assertTrue(output.getTable().getTableName() + " has a location which has a " +
+    for (ReadEntity input : hookContext.getInputs()) {
```
Comment (kgyrtkirk) [inline itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOutputTableLocationSchemeIsFileHook.java:27]: the logic here seem to have changed ; and its not aligned anymore with the name of the class

is this change intended? maybe you wanted to add a new hook which check for the input?

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java [hunk 2]
```diff
@@ -269,7 +269,19 @@ private static boolean isOwner(IMetaStoreClient metastoreClient, String userName
         thriftTableObj = metastoreClient.getTable(hivePrivObject.getDbname(),
             hivePrivObject.getObjectName());
       } catch (Exception e) {
-        throwGetObjErr(e, hivePrivObject);
+        boolean isTableExists = true;
+        try {
+          isTableExists = metastoreClient.tableExists(hivePrivObject.getDbname(), hivePrivObject.getObjectName());
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java:274]: I think you don't need the `isTableExists` boolean ; you could just:
```
if(!metastoreClient.tableExists(hivePrivObject.getDbname(), hivePrivObject.getObjectName())) {
  return true;
}
```

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java [hunk 3]
```diff
@@ -269,7 +269,19 @@ private static boolean isOwner(IMetaStoreClient metastoreClient, String userName
...
+        } catch (TException ex){
+          throwGetObjErr(ex, hivePrivObject);
+        }
+        if(!isTableExists){
+          // Do not throw any exception when table object is not present.
+          // return true since the current user the owner of new table.
+          return true;
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java:281]: I think instead of returning `true` here for non-existent tables; the "call-site" should check if the table already exists and make a decision based on that

File: ql/src/test/results/clientnegative/alter_non_native.q.out [hunk 4]
```diff
@@ -8,4 +8,15 @@ STORED BY 'org.apache.hadoop.hive.ql.metadata.DefaultStorageHandler'
...
+PREHOOK: type: ALTERTABLE_RENAME
+PREHOOK: Input: default@non_native1
+PREHOOK: Output: database:default
+PREHOOK: Output: default@new_non_native
+POSTHOOK: query: ALTER TABLE non_native1 RENAME TO new_non_native
+POSTHOOK: type: ALTERTABLE_RENAME
+POSTHOOK: Input: default@non_native1
+POSTHOOK: Output: database:default
+POSTHOOK: Output: default@new_non_native
+FAILED: AssertionError java.lang.AssertionError: Client Execution was expected to fail, but succeeded with error code 0 for fname=alter_non_native.q
```
Comment (kgyrtkirk) [inline ql/src/test/results/clientnegative/alter_non_native.q.out:21]: negative test is failing

File: ql/src/test/results/clientnegative/alter_table_wrong_db.q.out [hunk 5]
```diff
@@ -18,8 +18,4 @@ POSTHOOK: query: create table rename1(a int)
 POSTHOOK: type: CREATETABLE
 POSTHOOK: Output: bad_rename1@rename1
 POSTHOOK: Output: database:bad_rename1
-PREHOOK: query: alter table bad_rename1.rename1 rename to bad_db_notexists.rename1
-PREHOOK: type: ALTERTABLE_RENAME
-PREHOOK: Input: bad_rename1@rename1
-PREHOOK: Output: bad_rename1@rename1
-FAILED: Execution Error, return code 40013 from org.apache.hadoop.hive.ql.ddl.DDLTask. Unable to alter table. Unable to change partition or table. Object database hive.bad_db_notexists does not exist. Check metastore logs for detailed stack.
+FAILED: SemanticException [Error 10072]: Database does not exist: bad_db_notexists
```
Comment (kgyrtkirk) [inline ql/src/test/results/clientnegative/alter_table_wrong_db.q.out:21]: negative test fails with a different error

File: ql/src/test/results/clientnegative/alter_view_failure8.q.out [hunk 6]
```diff
@@ -6,4 +6,15 @@ POSTHOOK: query: CREATE TABLE invites (foo INT, bar STRING) PARTITIONED BY (ds S
...
+PREHOOK: type: ALTERVIEW_RENAME
+PREHOOK: Input: default@invites
+PREHOOK: Output: database:default
+PREHOOK: Output: default@invites2
+POSTHOOK: query: ALTER VIEW invites RENAME TO invites2
+POSTHOOK: type: ALTERVIEW_RENAME
+POSTHOOK: Input: default@invites
+POSTHOOK: Output: database:default
+POSTHOOK: Output: default@invites2
+FAILED: AssertionError java.lang.AssertionError: Client Execution was expected to fail, but succeeded with error code 0 for fname=alter_view_failure8.q
```
Comment (kgyrtkirk) [inline ql/src/test/results/clientnegative/alter_view_failure8.q.out:19]: negative test is failing

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java [hunk 7]
```diff
@@ -48,7 +54,12 @@ protected void analyzeCommand(TableName tableName, Map<String, String> partition
     if (AcidUtils.isTransactionalTable(table)) {
       setAcidDdlDesc(desc);
     }
-    addInputsOutputsAlterTable(tableName, null, desc, desc.getType(), false);
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java:51]: I think we are throwing out a lot of logic by not calling this method - and that might be the cause of the sheer volume of failures/test output changes

Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java:51]: I tried to implement essential logic in that method here which is organizing input and output object lists. But I now added back the method and added the output list in a way that we would need it in Apache Ranger.

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java (hunk 2 above)
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java]: Ack

File: itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOutputTableLocationSchemeIsFileHook.java (hunk 1 above)
Comment (saihemanth-cloudera) [reply itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOutputTableLocationSchemeIsFileHook.java:27]: Yeah, this change is an intended one. This class is only used in rename_table_location.q file and it is used  to verify the table object and its location. Previously, output and input had the old table objects for authorization. Now, we would like to have the old table in the input list and the new database and table object in the output list.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java [hunk 8]
```diff
@@ -49,6 +55,13 @@ protected void analyzeCommand(TableName tableName, Map<String, String> partition
       setAcidDdlDesc(desc);
     }
     addInputsOutputsAlterTable(tableName, null, desc, desc.getType(), false);
+//    inputs.add(new ReadEntity(table));
+    outputs.clear();
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java:59]: doing `outputs.clear()` right after `addInputsOutputsAlterTable` is pretty strange

could you please put this code into that method - isn't there where it supposed to live?

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java [hunk 9]
```diff
@@ -271,7 +271,16 @@ private static boolean isOwner(IMetaStoreClient metastoreClient, String userName
         thriftTableObj = metastoreClient.getTable(hivePrivObject.getDbname(),
             hivePrivObject.getObjectName());
       } catch (Exception e) {
-        throwGetObjErr(e, hivePrivObject);
+        boolean isTableExists = true;
+        try {
+          if(!metastoreClient.tableExists(hivePrivObject.getDbname(), hivePrivObject.getObjectName())) {
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java:276]: it would be better to place this check at the call site-s to this method - so that the explanation will not be needed at all

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java [hunk 10]
```diff
@@ -3612,8 +3612,8 @@ private void testRenameTable(boolean blocking) throws Exception {
       txnMgr2.acquireLocks(driver2.getPlan(), ctx, null, false);
       locks = getLocks();
 
-      ShowLocksResponseElement checkLock = checkLock(LockType.EXCLUSIVE,
-        LockState.WAITING, "default", "tab_acid", null, locks);
+      ShowLocksResponseElement checkLock = checkLock(LockType.SHARED_READ,
```
Comment (kgyrtkirk) [inline ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:3615]: Why we don't need an exclusive lock during a rename?

* what if we are reading data from a table in some dag
* in the meantime someone renames the table

with `SHARED_READ` the table will be renamed instantly - which will cause the dag to fail; or not?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java (hunk 8 above)
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java:59]: Yeah, I had to do this here. Reason: When a user executes Alter table foo rename to otherdatabase.foo2; we never send the otherdatabase and new table name (foo2) objects for authorization. We need to check whether the end-user has privileges to create a table in otherdatabase. So I think this special logic has to be in AbstractTableRenameAnalyzer instead of AbstractTableAnalyzer.

Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java:59]: I see - but this will make it more diverse...what do you think about the following:
* split the `addInputsOutputsAlterTable` into 2 methods which are taking care of inputs/outputs
* you could add back `addInputsOutputsAlterTable` which is calling these 2 for backward compatibility
* and call these 2 methods with old/new args from here?

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java (hunk 9 above)
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java:276]: I think this method should throw an exception if the table doesn't exists - you can't own non-exiting items; that's misleading...

* this is a private method - so it could only be called from this class
* the `metastoreClient` is passed as an argument to this method - so I don't think that should be a problem...

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java (hunk 10 above)
Comment (kgyrtkirk) [reply ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:3615]: this is the most serious issue - without addressing this I'm against this change

**Review by kgyrtkirk — CHANGES_REQUESTED** [review]

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java [hunk 11]
```diff
@@ -3754,7 +3754,7 @@ private void testRenameTable(boolean blocking) throws Exception {
     }
     driver2.lockAndRespond();
     locks = getLocks();
-    Assert.assertEquals("Unexpected lock count", 1, locks.size());
+    Assert.assertEquals("Unexpected lock count", 2, locks.size());
```
Comment (deniskuzZ) [inline ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:3757]: what's the other lock we are expecting here, should we check this in a test?

Comment (saihemanth-cloudera) [reply ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:3757]: one lock on the source table and another one on the destination.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java [hunk 12]
```diff
@@ -49,6 +55,11 @@ protected void analyzeCommand(TableName tableName, Map<String, String> partition
       setAcidDdlDesc(desc);
     }
     addInputsOutputsAlterTable(tableName, null, desc, desc.getType(), false);
+    String newDatabaseName = target.getDb() != null ? target.getDb() : table.getDbName(); // extract new database name from new table name, if not specified, then src dbname is used
+    Database newDatabase = getDatabase(newDatabaseName);
+    outputs.add(new WriteEntity(newDatabase, WriteEntity.WriteType.DDL_NO_LOCK));
+    Table newTable = new Table(target.getDb(), target.getTable());
+    outputs.add(new WriteEntity(newTable, WriteEntity.WriteType.DDL_EXCLUSIVE));
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java]: addInputsOutputsAlterTable(...) only adds lock on source table. we need to add locks on the destination as well. I have updated the patch to include a shared lock on the database and no lock on the newly created table object.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java [hunk 13]
```diff
@@ -49,6 +55,11 @@ protected void analyzeCommand(TableName tableName, Map<String, String> partition
       setAcidDdlDesc(desc);
     }
     addInputsOutputsAlterTable(tableName, null, desc, desc.getType(), false);
+    String newDatabaseName = target.getDb() != null ? target.getDb() : table.getDbName(); // extract new database name from new table name, if not specified, then src dbname is used
+    Database newDatabase = getDatabase(newDatabaseName);
+    outputs.add(new WriteEntity(newDatabase, WriteEntity.WriteType.DDL_SHARED));
+    Table newTable = new Table(target.getDb(), target.getTable());
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java:61]: I am not 100% sure, but shouldn't we propagate back from original table it's parameters, type, temp attribute, etc. See `DDLUtils.addDbAndTableToOutputs` & `SemanticAnalyzer.addDbAndTabToOutputs`

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java [hunk 14]
```diff
@@ -211,8 +211,15 @@ static RequiredPrivileges getPrivilegesFromMetaStore(IMetaStoreClient metastoreC
...
 
     // add owner privilege if user is owner of the object
-    if (isOwner(metastoreClient, userName, curRoles, hivePrivObject)) {
-      privs.addPrivilege(SQLPrivTypeGrant.OWNER_PRIV);
+    try {
+      if (metastoreClient.tableExists(hivePrivObject.getDbname(), hivePrivObject.getObjectName()) &&
+              isOwner(metastoreClient, userName, curRoles, hivePrivObject)) {
+        privs.addPrivilege(SQLPrivTypeGrant.OWNER_PRIV);
+      }
+    } catch (MetaException e) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java:219]: You could pipe multiple exceptions, i.e 
````
catch (MetaException | TException e) {
      throwGetPrivErr(e, hivePrivObject, userName);
}

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java [hunk 15]
```diff
@@ -211,8 +211,15 @@ static RequiredPrivileges getPrivilegesFromMetaStore(IMetaStoreClient metastoreC
     RequiredPrivileges privs = getRequiredPrivsFromThrift(thrifPrivs);
 
     // add owner privilege if user is owner of the object
-    if (isOwner(metastoreClient, userName, curRoles, hivePrivObject)) {
-      privs.addPrivilege(SQLPrivTypeGrant.OWNER_PRIV);
+    try {
+      if (metastoreClient.tableExists(hivePrivObject.getDbname(), hivePrivObject.getObjectName()) &&
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java:215]: PrivilageObjectType here might not be a TABLE_OR_VIEW.
````
    try {
      if (HivePrivilegeObjectType.TABLE_OR_VIEW != hivePrivObject.getType() ||
        metastoreClient.tableExists(hivePrivObject.getDbname(), hivePrivObject.getObjectName())) {

        if (isOwner(metastoreClient, userName, curRoles, hivePrivObject)) {
          privs.addPrivilege(SQLPrivTypeGrant.OWNER_PRIV);
        }
      }
    } catch (TException e) {
      throwGetPrivErr(e, hivePrivObject, userName);
    }
````

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java (hunk 11 above)
Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:3757]: please add the corresponding check in a test

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java (hunk 15 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java:215]: Wait, don't you need to add the privilege when the table DOESN'T exist? Also won't it hide the problem with the non-existing table for other operations besides RENAME??

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java (hunk 13 above)
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java:61]: That's a good point. I'll add them.

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java (hunk 15 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java:215]: - didn't you return 'true' in isOwner() when table doesn't exist in your initial patch and set the owner privilege?

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java (hunk 11 above)
Comment (saihemanth-cloudera) [reply ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:3757]: ack

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java (hunk 15 above)
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java]: ack

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java (hunk 13 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java:61]: why not just call
```` 
DDLUtils.addDbAndTableToOutputs(
        newDatabase, newTable
        table.getTableType(), table.isTemporary(), table.getParameters(), 
        outputs)`
````

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java [hunk 16]
```diff
@@ -49,6 +55,14 @@ protected void analyzeCommand(TableName tableName, Map<String, String> partition
...
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java:61]: should be 
Table newTable = new Table(newDatabaseName, target.getTable());

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLStdHiveAuthorizationValidator.java [hunk 17]
```diff
@@ -80,15 +80,15 @@ public void checkPrivileges(HiveOperationType hiveOpType, List<HivePrivilegeObje
 
     // check privileges on input and output objects
     List<String> deniedMessages = new ArrayList<String>();
-    checkPrivileges(hiveOpType, inputHObjs, metastoreClient, userName, IOType.INPUT, deniedMessages);
-    checkPrivileges(hiveOpType, outputHObjs, metastoreClient, userName, IOType.OUTPUT, deniedMessages);
+    checkPrivileges(hiveOpType, inputHObjs, metastoreClient, userName, IOType.INPUT, deniedMessages, false);
+    checkPrivileges(hiveOpType, outputHObjs, metastoreClient, userName, IOType.OUTPUT, deniedMessages, hiveOpType == HiveOperationType.ALTERTABLE_RENAME ? true: false);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLStdHiveAuthorizationValidator.java:84]: no need for a ternary operator here, just:
````
hiveOpType == HiveOperationType.ALTERTABLE_RENAME
````

File: itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOutputTableLocationSchemeIsFileHook.java (hunk 1 above)
Comment (deniskuzZ) [inline itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOutputTableLocationSchemeIsFileHook.java:27]: Why is this change? switch from outputs -> inputs?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java (hunk 13 above)
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/rename/AbstractAlterTableRenameAnalyzer.java]: ack

File: itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOutputTableLocationSchemeIsFileHook.java (hunk 1 above)
Comment (saihemanth-cloudera) [reply itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOutputTableLocationSchemeIsFileHook.java]: I have updated the patch to verify the output table i.e..., the old table which is the desired condition.
